egs2/librispeech/discrete_asr1/local/data.sh

```diff
@@ -0,0 +1,452 @@
+if [ ${stage} -le 3 ] && [ ${stop_stage} -ge 3 ] && ! [[ " ${skip_stages} " =~ [[:space:]]3[[:space:]] ]]; then
+    log "Stage 3: dump features"
+
+    # download ssl ckpt
+    mkdir -p "./ckpt"
+
+    download_model_ckpts "${model_type}" "${model_version}"
+    if [ ! -f "./ckpt/${ckpt_name}" ]; then
+        wget -O "./ckpt/${ckpt_name}" ${model_url}
+    fi
+
+    for dset in ${test_sets} "${train_set}" ; do
+        log "Dump SSL ${dset} features to ${feats_dir}/${dset}"
+
+        # 1. Split the key file
+        utils/copy_data_dir.sh --validate_opts --non-print data/dump_audio/"${dset}" "${feats_dir}/${dset}"
+        rm -f "${feats_dir}/${dset}"/{segments,wav.scp,reco2file_and_channel,reco2dur}
+
+        output_dir="${feats_dir}/${dset}/data"
+        mkdir -p "${output_dir}"
+        _logdir="${feats_dir}/${dset}/logdir"
+        mkdir -p "${_logdir}"
+
+        nutt=$(<"data/dump_audio/${dset}"/wav.scp wc -l)
+        _nj=$((gpu_nj<nutt?gpu_nj:nutt))
+
+        key_file="data/dump_audio/${dset}"/wav.scp
+        split_scps=""
+        for n in $(seq ${_nj}); do
+            split_scps+=" ${_logdir}/wav.${n}.scp"
+        done
+        # shellcheck disable=SC2086
+        utils/split_scp.pl "${key_file}" ${split_scps}
+
+        ${cuda_cmd} --gpu 1 JOB=1:${_nj} ${_logdir}/dump_features.JOB.log \
+            ${python} local/dump_hubert_or_wavlm_feature.py \
+                --ckpt_path "./ckpt/${ckpt_name}" \
+                --layer ${layer} \
+                --ssl_type "${model_type}" \
+                --in_filetype "sound" \
+                --out_filetype "mat" \
+                --write_num_frames "ark,t:${output_dir}/utt2num_frames.JOB" \
+                "scp:${_logdir}/wav.JOB.scp" \
+                "ark,scp:${output_dir}/feats.JOB.ark,${output_dir}/feats.JOB.scp" || exit 1;
+
+        # concatenate scp files
+        for n in $(seq ${_nj}); do
+            cat ${output_dir}/feats.${n}.scp || exit 1;
+        done > ${output_dir}/../feats.scp || exit 1
+
+        for n in $(seq ${_nj}); do
+            cat ${output_dir}/utt2num_frames.$n || exit 1;
+        done > ${output_dir}/../utt2num_frames || exit 1
+
+    done
+
+fi
+
+km_dir=exp/kmeans/$(echo "${feats_type}" | tr "+" "_")_${nclusters}clusters
+if [ ${stage} -le 4 ] && [ ${stop_stage} -ge 4 ] && ! [[ " ${skip_stages} " =~ [[:space:]]4[[:space:]] ]]; then
+    log "Stage 4: learn kmeans"
+    mkdir -p ${km_dir}
+
+    # select portion of data
+    nutt=$(<"${feats_dir}/${train_set}"/feats.scp wc -l)
+    portion_nutt=$(echo ${nutt} ${portion} | awk '{print(int($1 * $2)+1)}')
+
+    subset_scp.pl ${portion_nutt} "${feats_dir}/${train_set}"/feats.scp \
+        > "${km_dir}/train.scp" || exit 1;
+    log "Subsampling ${portion_nutt} utterances from ${train_set} for Kmeans training."
+
+    # It typically requires 120GB RAM to run kmeans steps.
+    ${train_cmd} --num_threads 12 --mem 15G ${km_dir}/logdir/learn_kmeans.log \
```

Comment:
Do we have the distributed version for kmeans in espnet now?

Reply:
I think it still in progress. When we update it in the hubert recipe, it can be applied here in the same way.